examples/image-telephone/application.py

```diff
@@ -0,0 +1,170 @@
+"""
+This module demonstrates a telephone application
+using Burr that:
+ - captions an image
+ - creates caption embeddings (for analysis)
+ - creates a new image based on the created caption
+"""
+import os
+import uuid
+
+import requests
+from hamilton import dataflows, driver
+
+from burr.core import Action, ApplicationBuilder, State, default, expr
+from burr.core.action import action
+from burr.lifecycle import PostRunStepHook
+
+# import hamilton modules
+caption_images = dataflows.import_module("caption_images", "elijahbenizzy")
+generate_images = dataflows.import_module("generate_images", "elijahbenizzy")
+
+
+class ImageSaverHook(PostRunStepHook):
+    """Class to save images to a directory.
+    This is an example of a custom way to interact indirectly.
+
+    This is one way you could save to S3 by writing something like this.
+    """
+
+    def __init__(self, save_dir: str = "saved_images"):
+        self.save_dir = save_dir
+        self.run_id = str(uuid.uuid4())[0:8]
+        self.path = os.path.join(self.save_dir, self.run_id)
+        if not os.path.exists(self.path):
+            os.makedirs(self.path)
+
+    def post_run_step(self, *, state: "State", action: "Action", **future_kwargs):
+        """Pulls the image URL from the state and saves it to the save directory."""
+        if action.name == "generate":
+            image_url = state["current_image_location"]
+            image_name = "image_" + str(state["__SEQUENCE_ID"]) + ".png"
+            with open(os.path.join(self.path, image_name), "wb") as f:
+                f.write(requests.get(image_url).content)
+                print(f"Saved image to {self.path}/{image_name}")
+
+
+@action(
+    reads=["current_image_location"],
+    writes=["current_image_caption", "image_location_history"],
+)
+def image_caption(state: State, caption_image_driver: driver.Driver) -> tuple[dict, State]:
```

Comment:
We're not using the `Hamilton` abstraction -- this is going to confuse people. MIght be worth deleting that or fixing it up to include `overrides` if we don't like it, otherwise we should not have two ways to do things.

Reply:
Ahh I see one that uses it. Let's make it clear that we have both of these.